File: mcumgr-core/src/test/java/io/runtime/mcumgr/McuMgrImageTest.java
package io.runtime.mcumgr;

import org.junit.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

import io.runtime.mcumgr.image.McuMgrImage;

import static org.junit.Assert.assertNotNull;

public class McuMgrImageTest {

    @Test
    public void fromBytes_unprotectedTlvs_success() throws Exception {
        ClassLoader classLoader = this.getClass().getClassLoader();
        assertNotNull(classLoader);
        InputStream inputStream = classLoader.getResourceAsStream("slinky-no-prot-tlv.img");
        assertNotNull(inputStream);
        byte[] imageData = toByteArray(inputStream);
        McuMgrImage.fromBytes(imageData);
    }

    @Test
    public void fromBytes_protectedTlvs_success() throws Exception {
        ClassLoader classLoader = this.getClass().getClassLoader();
        assertNotNull(classLoader);
        InputStream inputStream = classLoader.getResourceAsStream("slinky-prot-tlv.img");
        assertNotNull(inputStream);
        byte[] imageData = toByteArray(inputStream);
        McuMgrImage.fromBytes(imageData);
    }

    private static byte[] toByteArray(InputStream in) throws IOException {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        byte[] buffer = new byte[1024];
        int len;
        while ((len = in.read(buffer)) != -1) {
            os.write(buffer, 0, len);
        }
        return os.toByteArray();
    }
}

File: mcumgr-core/src/test/java/io/runtime/mcumgr/McuMgrImageTest.kt
package io.runtime.mcumgr

import io.runtime.mcumgr.image.McuMgrImage
import org.junit.Test
import java.io.ByteArrayOutputStream
import java.io.InputStream

class McuMgrImageTest {

    @Test
    fun `parse image without protected tlvs success`() {
        val inputStream = this::class.java.classLoader?.getResourceAsStream("slinky-no-prot-tlv.img")!!
            ?: throw IllegalStateException("input stream is null")
        val imageData = toByteArray(inputStream)
        McuMgrImage.fromBytes(imageData)
    }

    @Test
    fun `parse image with protected tlvs success`() {
        val inputStream = this::class.java.classLoader?.getResourceAsStream("slinky-prot-tlv.img")
            ?: throw IllegalStateException("input stream is null")
        val imageData = toByteArray(inputStream)
        McuMgrImage.fromBytes(imageData)
    }

    private fun toByteArray(inputStream: InputStream): ByteArray {
        val os = ByteArrayOutputStream()
        val buffer = ByteArray(1024)
        var len: Int
        while (inputStream.read(buffer).also { len = it } != -1) {
            os.write(buffer, 0, len)
        }
        return os.toByteArray()
    }
}
